File: cloud_functions/transcribe_audio_save_db/services/secret_manager.py
"""
Module for interacting with Google Cloud Secret Manager.
"""

import os
import logging
from google.cloud import secretmanager
from google.api_core.exceptions import NotFound, PermissionDenied

from config import is_production

# Initialize logger
logger = logging.getLogger(__name__)

# Global variables
_secret_client = None


def get_secret_client():
    """Get or create Secret Manager client"""
    global _secret_client

    if _secret_client is None:
        _secret_client = secretmanager.SecretManagerServiceClient()

    return _secret_client


def access_secret(project_id, secret_name, version="latest"):
    """Access a secret from Secret Manager"""
    if not is_production():
        logger.info(
            f"Not in production, skipping Secret Manager access for {secret_name}"
        )
        return None

    try:
        client = get_secret_client()

        # Build the resource name
        if version == "latest":
            name = f"projects/{project_id}/secrets/{secret_name}/versions/latest"
        else:
            name = f"projects/{project_id}/secrets/{secret_name}/versions/{version}"

        # Access the secret version
        response = client.access_secret_version(request={"name": name})

        # Return the decoded payload
        return response.payload.data.decode("UTF-8")

    except NotFound:
        logger.warning(f"Secret {secret_name} not found")
        return None
    except PermissionDenied:
        logger.error(f"Permission denied accessing secret {secret_name}")
        return None
    except Exception as e:
        logger.error(f"Error accessing secret {secret_name}: {str(e)}")
        return None


def load_environment_from_secret_manager(project_id=None):
    """Load environment variables from Secret Manager in production"""
    if not is_production():
        logger.info("Not in production, skipping Secret Manager configuration")
        return False

    try:
        project_id = project_id or os.environ.get("PROJECT_ID")
        if not project_id:
            logger.error("PROJECT_ID not set, cannot load secrets")
            return False

        logger.info("Loading configuration from Secret Manager")

        # Load key configuration values from Secret Manager
        # Format: secret name -> environment variable name
        secrets_to_load = {
            "gemini-api-key": "GEMINI_API_KEY",
            "gemini-model": "GEMINI_MODEL",
            "temperature": "TEMPERATURE",
            "top-p": "TOP_P",
            "max-output-tokens": "MAX_OUTPUT_TOKENS",
        }

        # Load each secret and set as environment variable
        for secret_name, env_var in secrets_to_load.items():
            secret_value = access_secret(project_id, secret_name)
            if secret_value:
                os.environ[env_var] = secret_value
                logger.info(f"Loaded {env_var} from Secret Manager")
            else:
                logger.warning(f"Failed to load {env_var} from Secret Manager")

        return True

    except Exception as e:
        logger.error(f"Error loading secrets: {str(e)}")
        return False

File: next_app/src/lib/secretManager.ts
import { SecretManagerServiceClient } from "@google-cloud/secret-manager";

// This will only run on the server side
export async function getSecret(secretName: string): Promise<string> {
    // Use local env variables during development
    if (process.env.NODE_ENV === "development") {
        return process.env[secretName] || "";
    }

    try {
        // Get environment suffix for test or production
        const environment = process.env.ENVIRONMENT || "test"; // Default to test if not specified
        const client = new SecretManagerServiceClient();
        const projectId = process.env.GCP_PROJECT_ID;

        if (!projectId) {
            console.error("GCP_PROJECT_ID environment variable is not set");
            return "";
        }

        // Format the secret name with environment suffix for different environments
        // For example: NEXT_PUBLIC_API_URL_test or NEXT_PUBLIC_API_URL_production
        const environmentSpecificName = `${secretName}_${environment}`;

        try {
            // First try to get environment-specific secret
            const [version] = await client.accessSecretVersion({
                name: `projects/${projectId}/secrets/${environmentSpecificName}/versions/latest`,
            });
            const payload = version.payload?.data?.toString() || "";
            return payload;
        } catch (error) {
            console.log(
                `No environment-specific secret found for ${environmentSpecificName}, trying generic version`
            );

            // Fall back to generic secret name
            const [version] = await client.accessSecretVersion({
                name: `projects/${projectId}/secrets/${secretName}/versions/latest`,
            });
            const payload = version.payload?.data?.toString() || "";
            return payload;
        }
    } catch (error) {
        console.error(`Error fetching secret ${secretName}:`, error);
        return "";
    }
}

// Load all secrets at once
export async function loadSecrets(): Promise<void> {
    if (process.env.NODE_ENV === "development") {
        console.log("Running in development mode, using local .env files");
        return;
    }

    const environment = process.env.ENVIRONMENT || "test";
    console.log(`Loading secrets for ${environment} environment`);

    const secrets = [
        "NEXT_PUBLIC_API_URL",
        // Add other secrets as needed
    ];

    console.log(
        `Loading secrets from Google Cloud Secret Manager for ${environment} environment`
    );

    try {
        await Promise.all(
            secrets.map(async (secretName) => {
                const value = await getSecret(secretName);
                if (value) {
                    process.env[secretName] = value;
                } else {
                    console.warn(`No value found for secret: ${secretName}`);
                }
            })
        );
        console.log(
            "Successfully loaded secrets from Google Cloud Secret Manager"
        );
    } catch (error) {
        console.error("Failed to load secrets:", error);
    }
}
